### PR TITLE
Compiling OpenMC using Visual Studio

### LINKED ISSUE #1248: Get rid of variable length arrays (VLAs) in C++
As observed in #1243, trying to compile with MSVC results in errors due to the use of variable-length arrays, which are not part of the C++ standard. We've been getting by because evidently g++, clang++, and icpc all support VLAs as an extension. The simplest alternative is to use a heap-allocated vector, i.e., instead of
```C++
void foo(int n) {
  T bar[n];
}
```
you could have
```C++
void foo(int n) {
  std::vector<T> bar(n);
}
```
One potential concern of mine is whether any VLAs are used during transport (i.e., within OpenMP threads). If we switch to std::vector, there could be possible thread contention on a mutex for `malloc`.

### PR DESCRIPTION
**Pull request by liangjg**
This PR tries to make a step towards [windows support #1243](https://github.com/openmc-dev/openmc/issues/1243): building openmc on windows with [Visual Studio](https://visualstudio.microsoft.com/zh-hans/?rr=https%3A%2F%2Fwww.google.com%2F).
In this approach, the Visual Studio solution can be generated using CMake with the existing `CMakeLists.txt`, with the aid of [CMake-generators](https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html).

``` bash
git clone https://github.com/openmc-dev/openmc.git
cd openmc

mkdir build_vs
cd build_vs
cmake -G "Visual Studio 16 2019" -A x64 # cmake -G --help to see a list of generators

# then, just open "openmc.sln" and build openmc using Visual Studio 
```
The workflow is preliminarily tested on my windows 10 and with Visual Studio 2019 community IDE. I managed to run a simple pincell problem. But more testing is needed.

Some notes:
- Variable length arrays (VLAs) are replaced for VC++ compatiblity, to close #1248.
- [To use the shared HDF5 libraries on Windows with Visual Studio you must specify the H5_BUILT_AS_DYNAMIC_LIB compile definition](https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.8/hdf5-1.8.16/obtain51816.html).
- `libopenmc` library is built as `STATIC` instead of `SHARED` for Visual Studio since you must explicitly [exporting symbols for dynamic library](https://docs.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=vs-2019) linking in Visual Studio, which means the code needs to be changed.
- Currently building with `openmp` is not working because of the following compiling issue, which seems [a violation of VC++ for openmp standard](https://stackoverflow.com/questions/12560243/using-threadprivate-directive-in-visual-studio).
``` C++
extern std::vector<Particle::Bank> fission_bank;
#pragma omp threadprivate(fission_bank)

error C3053: 'fission_bank' : 'threadprivate' is only valid for global or static data items.
```

### CODE REVIEW

**Comment by pshriwise** [comment]
I also had success in getting in building with MSVC 17. The only change I had to make was to add `include <cctype>` in `cell.cpp`.

File: src/surface.cpp [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <cmath>
 #include <sstream>
 #include <utility>
+#include <ciso646>
```
[the remark this answers is not in the document]
Comment (liangjg) [reply src/surface.cpp]: ok

File: src/tallies/filter_legendre.cpp [hunk 2]
```diff
@@ -18,8 +18,8 @@ void
 LegendreFilter::get_all_bins(const Particle* p, int estimator,
                              FilterMatch& match) const
 {
-  double wgt[n_bins_];
-  calc_pn_c(order_, p->mu_, wgt);
+  std::vector<double> wgt(n_bins_);
+  calc_pn_c(order_, p->mu_, &wgt[0]);
```
[the remark this answers is not in the document]
Comment (liangjg) [reply src/tallies/filter_legendre.cpp]: ok

File: src/initialize.cpp [hunk 3]
```diff
@@ -68,7 +68,7 @@ int openmc_init(int argc, char* argv[], const void* intracomm)
 
   // Initialize random number generator -- if the user specifies a seed, it
   // will be re-initialized later
-  openmc_set_seed(DEFAULT_SEED);
+  openmc::openmc_set_seed(DEFAULT_SEED);
```
[the remark this answers is not in the document]
Comment (liangjg) [reply src/initialize.cpp]: My VS says it is an ambiguous call without `openmc::`
```
\src\initialize.cpp(71,31): error C2668:  'openmc_set_seed': ambiguous call to overloaded function
include\openmc/capi.h(78,8): message :  could be 'void openmc_set_seed(int64_t)'
\include\openmc/random_lcg.h(94,17): message :  or       'void openmc::openmc_set_seed(int64_t)'
src\initialize.cpp(71,31): message :  while trying to match the argument list '(const int64_t)
```

**Comment by liangjg** [comment]
@paulromano I installed HDF5 using the [Pre-built Binary Distributions for windows](https://www.hdfgroup.org/downloads/hdf5/) and CMake found it automatically.

**Comment by liangjg** [comment]
@pshriwise I added the `include`, though not sure why this error didn't show up in VS2019..

**Pull request merged**